### ISSUE TITLE
Loss isn't decreasing

### ISSUE THREAD

**Issue by IamAdiSri** [tebesu/CollaborativeMemoryNetwork#5]
Hi, I'm trying to re-implement this project in PyTorch. I've written the model following the Tensorflow implementation given here, however, my loss does not converge.

I believe I might be getting the trainable parameters wrong. Could you please tell me which parameters are trainable in your model?

If it's not too much trouble could you also have a look at my code?
[Link](https://github.com/IamAdiSri/cmn4recosys/blob/master/CMN%20Pytorch.ipynb)

**Comment by tebesu** [comment]
From looking at the code, I did not run it but it seems you are applying the L2 regularization twice.

The first one you apply is in the optimizer
`optimizer = torch.optim.RMSprop(model.parameters(), lr=config.learning_rate, 
                                momentum=config.momentum, weight_decay=config.decay_rate)`

https://pytorch.org/docs/stable/optim.html#torch.optim.RMSprop

The parameter to `weight_decay` is the L2 regularization not decay rate. Then you add another l2 in the loss criterion and here


`        batch_loss = criterion(score_pos, score_neg)
        
        # adding l2 regularisation
        for name, param in model.named_parameters():
            if name in ['mem_layer.hop_mapping.weight', 
                        'output_module.dense.weight', 
                        'output_module.out.weight']:
                l2 = torch.sqrt(param.pow(2).sum())
                batch_loss += (config.l2_lambda * l2)`

Try removing it all together and see if it converges.

**Comment by IamAdiSri** [comment]
Thank you for taking the time to look through my code. 

I changed the **weight_decay** parameter from **decay_rate (0.9)** to **l2_lamba (0.1)** penalty in the optimizer.

`optimizer = torch.optim.RMSprop(model.parameters(), lr=config.learning_rate, momentum=config.momentum, weight_decay=config.l2_lambda)`

I also removed the manual implementation for the l2 regularization you've mentioned above.

Lastly, I also removed the l2_regularization when calculating Bayesian Personalized Loss.

On running the model again, the loss has decreased from the previous ~.76, but stagnates again at ~0.69; the model still doesn't converge. Do you have any other suggestions?

**Comment by IamAdiSri** [comment]
@tebesu Removing loss entirely from the optimizer and then performing loss calculation manually, somehow made it work. The model now trains and loss decreases.

Thank you for your help.

**Comment by tebesu** [comment]
@IamAdiSri glad you were able to get it to work, let me know if you have additional questions I will do my best to assist you.
